Implementor feedback from mls-extensions

rohanmahy opened the pull request:
- AppDataUpdate and SelfRemove used the same proposal number. SelfRemove -> 0x000a
- Remove the last references to safe extensions structs in TargetedMessage
- targeted_messages_capability is no longer needed with supported|required_wire_formats
- Register missing MLS Signature Labels to IANA section
- Add GREASE values for components

I could go further by adding GREASE for wire_format (in capabilities only).

Commit:
Make SelfRemove proposal use a unused proposal number (0x000a)

Commit:
TargetedMessage is no longer a "safe extension", it is just a regular extension. Replumb appropriately

Commit:
targeted_messages_capability is no longer needed since there is a generic supported and required wire formats extension

Commit:
Add MLS Signature Labels from TargetedMessage and multi-credentials

Commit:
Add GREASE values for components

Commit:
Update change log

Commit:
Remove text from old safe extension framework

raphaelrobert commented on draft-ietf-mls-extensions.md:
```diff
@@ -830,7 +830,8 @@ used. In pseudocode, the key and nonce are derived as:
 sender_auth_data_secret
-  = DeriveExtensionSecret(extension_secret, "targeted message sender auth data")
+  = DeriveSecret(epoch_secret,
+                    "targeted message sender auth data secret")
```
I don't think this is necessary. We could just use the normal exporter, since we don't need FS anyway. That would make it much simpler to implement.

rohanmahy replied:
@raphaelrobert ping ^^^

raphaelrobert commented on draft-ietf-mls-extensions.md:
```diff
@@ -881,14 +882,13 @@ described in {{safe-hpke}} with an empty context. Other functions are defined in
 `TargetedMessageAuthScheme.HPKEAuthPsk`.
 
-As described in {{safe-hpke}} the `hpke_context` is a LabeledExtensionContent struct
+As described in {{safe-hpke}} the `hpke_context` is an ExtensionContext struct
```
I'm not sure what an `ExtensionContext` struct is. Is that defined anywhere?

I think this whole section doesn't make sense anymore. The HPKE section has changed, but also seems incomplete.

bifurcation replied:
@raphaelrobert Are you saying the HPKE section seems incomplete?  Because it only covers the Base mode?  Seems like a discussion for another issue.

I agree that this section needs to either get removed or be refactored to refer to the safe HPKE interface as it is now defined.  I would probably argue for the former, since Auth/AuthPSK modes don't work with any PQ KEMs.

raphaelrobert replied:
I hadn't realized the auth mode wasn't exposed. I'm fine dropping it, it would streamline targeted messages. 
The fact that the AuthPSK mode is not exposed is however problematic. It was used for channel binding, so that targeted messages required the recipient to be a full group member and not just the owner of the private key that corresponds to HPKE public key of a leaf.

For the sake of progress, I propose to revert the change above and do a separate PR that aligns targeted messages with the safe API, provided we find a solution to the PSK injection.

raphaelrobert commented on draft-ietf-mls-extensions.md:
```diff
@@ -1472,20 +1456,39 @@ from a group more efficiently than using a `remove` proposal type, as the
+As described in {{Section 13.5 of !RFC9420}}, GREASE values are ignored by
+receivers, while a random selection of GREASE values are included in certain
+fields. This specification adds the following locations where GREASE values for
+components can be included:
```
I think this paragraph should use normative language.

bifurcation replied:
I agree with with normative language.  I would suggest that GREASE SHOULD be added to these fields _if present_.  To emphasize that you don't need to add these fields just to send GREASE.

raphaelrobert reviewed (changes requested):
Thanks for tackling some of the issues. I think more GREASE values would be a good thing. I marked a few things I don't agree with/don't understand.

bifurcation commented on draft-ietf-mls-extensions.md:
```diff
@@ -830,7 +830,8 @@ used. In pseudocode, the key and nonce are derived as:
 ~~~ tls
 sender_auth_data_secret
-  = DeriveExtensionSecret(extension_secret, "targeted message sender auth data")
+  = DeriveSecret(epoch_secret,
+                    "targeted message sender auth data secret")
```
```suggestion
  = MLS-Exporter("targeted message", "sender auth data secret", KDF.Nh)
```

raphaelrobert replied:
Thanks, that's exactly what I meant

bifurcation commented on draft-ietf-mls-extensions.md:
```diff
@@ -869,7 +870,7 @@ For the PSK part of the authentication, clients export a dedicated secret:
 ~~~ tls
 targeted_message_psk
-  = DeriveExtensionSecret(extension_secret, "targeted message psk")
+  = DeriveSecret(epoch_secret, "targeted message psk")
```
MLS-Exporter("targeted message", "psk", KDF.Nh)

bifurcation commented on draft-ietf-mls-extensions.md:
```diff
@@ -958,11 +958,10 @@ message = OpenPSK(kem_output,
-                        "TargetedMessageTBS",
-                        targeted_message_tbs,
-                        signature)
+VerifyWithLabel.verify(sender_leaf_node.signature_key,
+                       "TargetedMessageTBS",
+                       targeted_message_tbs,
+                       signature)
```
This should be SafeVerifyWithLabel, we just need to change the `extension_type` to a `component_id`, presumably for some fixed component ID that identifies this targeted messages mechanism.

rohanmahy committed:
replace DeriveSecret() with MLS-Exporter()

Co-authored-by: Richard Barnes <rlb@ipv.sx>

Commit:
fix hpke_context definition in TargetedMessage

Commit:
move GREASE description under Negotiation; add normative language

rohanmahy commented:
I think I have addressed all comments received so far. @raphaelrobert @bifurcation please have a look.

I wish github had a "propose resolved" button.

raphaelrobert commented on draft-ietf-mls-extensions.md:
```diff
@@ -881,16 +906,18 @@ described in {{safe-hpke}} with an empty context. Other functions are defined in
+} TargetedMessageContext;
 
+label = "MLS 1.0 TargetedMessageData"
+context = group_context
```
This should likely change, but I'll do a follow-up PR. The group context is public, meaning you don't have to be a member to know it. I'll try to see if we can replace the AuthPSK mode with that, but I'm not sure yet.

Anyway, we can merge this, and I'll rewrite it later.

rohanmahy commented:
@raphaelrobert please go ahead and merge. I don't have permission to in this repo.

raphaelrobert commented:
> I wish github had a "propose resolved" button.

You can always just request a review, if that's what you meant.
